src/coordinates/uniform_cartesian.hpp

```diff
@@ -215,6 +249,32 @@ class UniformCartesian {
     return cell_volume_;
   }
 
+  //----------------------------------------
+  // Generalized volume
+  //----------------------------------------
+  template <TopologicalElement el, class... Args>
+  KOKKOS_FORCEINLINE_FUNCTION Real Volume(Args... args) const {
+    using TE = TopologicalElement;
+    if constexpr (el == TE::CC) {
+      return cell_volume_;
+    } else if constexpr (el == TE::F1) {
+      return area_[X1DIR - 1];
+    } else if constexpr (el == TE::F2) {
+      return area_[X2DIR - 1];
+    } else if constexpr (el == TE::F3) {
+      return area_[X3DIR - 1];
+    } else if constexpr (el == TE::E1) {
+      return dx_[X1DIR - 1];
+    } else if constexpr (el == TE::E2) {
+      return dx_[X2DIR - 1];
+    } else if constexpr (el == TE::E3) {
+      return dx_[X3DIR - 1];
+    } else if constexpr (el == TE::NN) {
+      return 1.0;
+    }
+    return 0.0;
```

Comment:
Maybe add a parthenon fail for this one?

Reply:
Possibly, but you should never get there because the if clauses exhaust all possible types of topological elements.

Reply:
Yeah I know, but I kind of want the fail for readability---a clear message "you should never reach this." Same for the above comment.